src/iop/channelmixer.c

```diff
@@ -146,92 +146,144 @@ void connect_key_accels(dt_iop_module_t *self)
+                           0.0f, 1.0f);
+      smix = clamp_range_f((rgb[0] * data->red[CHANNEL_SATURATION])
+                           + (rgb[1] * data->green[CHANNEL_SATURATION])
+                           + (rgb[2] * data->blue[CHANNEL_SATURATION]),
+                           0.0f, 1.0f);
+      lmix = clamp_range_f((rgb[0] * data->red[CHANNEL_LIGHTNESS])
+                           + (rgb[1] * data->green[CHANNEL_LIGHTNESS])
+                           + (rgb[2] * data->blue[CHANNEL_LIGHTNESS]),
+                           0.0f, 1.0f);
 
       // If HSL mix is used apply to out[]
       if(hmix != 0.0 || smix != 0.0 || lmix != 0.0)
       {
+        // rgb2hsl expects all values to be clipped
+        rgb[0] = clamp_range_f(rgb[0], 0.0f, 1.0f);
+        rgb[1] = clamp_range_f(rgb[1], 0.0f, 1.0f);
+        rgb[2] = clamp_range_f(rgb[2], 0.0f, 1.0f);
         // mix into HSL output channels
-        rgb2hsl(in, &h, &s, &l);
+        rgb2hsl(rgb, &h, &s, &l);
         h = (hmix != 0.0) ? hmix : h;
         s = (smix != 0.0) ? smix : s;
         l = (lmix != 0.0) ? lmix : l;
-        hsl2rgb(out, h, s, l);
-      }
-      else // no HSL copt in[] to out[]
-        for(int c = 0; c < 3; c++) out[c] = in[c];
-
-      // Calculate graymix and RGB mix
-      graymix = CLIP((out[0] * data->red[CHANNEL_GRAY]) + (out[1] * data->green[CHANNEL_GRAY])
-                     + (out[2] * data->blue[CHANNEL_GRAY]));
-
-      rmix = CLIP((out[0] * data->red[CHANNEL_RED]) + (out[1] * data->green[CHANNEL_RED])
-                  + (out[2] * data->blue[CHANNEL_RED]));
-      gmix = CLIP((out[0] * data->red[CHANNEL_GREEN]) + (out[1] * data->green[CHANNEL_GREEN])
-                  + (out[2] * data->blue[CHANNEL_GREEN]));
-      bmix = CLIP((out[0] * data->red[CHANNEL_BLUE]) + (out[1] * data->green[CHANNEL_BLUE])
-                  + (out[2] * data->blue[CHANNEL_BLUE]));
-
-
-      if(gray_mix_mode) // Graymix is used...
-        out[0] = out[1] = out[2] = graymix;
-      else // RGB mix is used...
-      {
-        out[0] = rmix;
-        out[1] = gmix;
-        out[2] = bmix;
+        hsl2rgb(rgb, h, s, l);
       }
+    }
 
-      /*mix = CLIP( in[0] * data->red)+( in[1] * data->green)+( in[2] * data->blue );
+    // Calculate RGB mix
+    rmix = (rgb[0] * data->red[CHANNEL_RED])
+           + (rgb[1] * data->green[CHANNEL_RED])
+           + (rgb[2] * data->blue[CHANNEL_RED]);
+    gmix = (rgb[0] * data->red[CHANNEL_GREEN])
+           + (rgb[1] * data->green[CHANNEL_GREEN])
+           + (rgb[2] * data->blue[CHANNEL_GREEN]);
+    bmix = (rgb[0] * data->red[CHANNEL_BLUE])
+           + (rgb[1] * data->green[CHANNEL_BLUE])
+           + (rgb[2] * data->blue[CHANNEL_BLUE]);
+
+    rgb[0] = rmix;
+    rgb[1] = gmix;
+    rgb[2] = bmix;
+
+    if(gray_mix_mode)
+    {
+      // Calculate graymix
+      // clip blacks as negative values could be amplified by channel factors
+      graymix = (clamp_range_f(rgb[0], 0.0f, INFINITY) * data->red[CHANNEL_GRAY])
```

Comment:
why clamp between `0.f` and `INFINITY` ? why not a simple `fmax(rgb, 0.f)` ? or did you mean to clamp between `0.f` and  `clip_value` ?

Reply:
Replaced this by fmax(rgb, 0.f), thanks